ykrt/src/compile/jitc_yk/gdb.rs

```diff
@@ -0,0 +1,241 @@
+// dereferencable.
+#[derive(Debug)]
+#[deku_derive(DekuWrite)]
+struct YkSymFile {
+    sym_name: CString,
+    jitted_code_vaddr: usize,
+    jitted_code_size: usize,
+    src_path: std::ffi::CString,
+    num_lineinfos: c_int,
+    #[deku(count = "num_lineinfos")]
+    lineinfos: Vec<LineInfo>,
+}
+
+/// The gdb context for a trace.
+///
+/// This contains resources that need to be kept-alive for the trace to be safely debuggable.
+#[derive(Debug)]
+pub(crate) struct GdbCtx {
+    /// The file containing the "source code" that we show in gdb.
+    #[allow(dead_code)]
+    src_file: tempfile::NamedTempFile,
+    /// The entry into gdb's linked list.
+    #[allow(dead_code)]
+    jit_code_entry: Box<JitCodeEntry>,
+    /// The serialised payload of the symbol file.
+    #[allow(dead_code)]
+    payload: Vec<u8>,
+}
+
+pub(crate) fn register_jitted_code(
+    id: u64,
+    jitted_code: *const u8,
+    jitted_code_size: usize,
+    comments: &IndexMap<usize, Vec<String>>,
+) -> GdbCtx {
+    // Write the comment lines out to a "source code file" that we want gdb to show lines from. As
+    // we do this, we also build a mapping from virtual addresses to line numbers.
+    let mut src_file = NamedTempFile::new().unwrap();
+    let mut lineinfos = Vec::new();
+    let mut line_num = 1;
+    let code_vaddr = jitted_code as usize;
+    for (off, lines) in comments {
+        lineinfos.push(LineInfo {
+            vaddr: code_vaddr + off,
+            line_num,
+        });
+        for line in lines {
+            writeln!(src_file, "{}", line).unwrap();
+            line_num += 1;
+        }
+    }
+
+    let sym_name = CString::new(format!("__yk_compiled_trace{}", id)).unwrap();
+
+    // Build the symbol file we are going to give to gdb.
+    let sym_file = Box::new(YkSymFile {
+        sym_name,
+        jitted_code_vaddr: jitted_code as usize, // cast safe: ptr and usize the same size.
+        jitted_code_size,
+        src_path: CString::new(src_file.path().to_str().unwrap()).unwrap(),
+        num_lineinfos: c_int::try_from(lineinfos.len()).unwrap(),
+        lineinfos,
+    });
+
+    // And serialise it for gdb to read.
+    //
+    // The serialised payload is conceptually an owned copy, so it's ok to let the original drop.
+    let payload = sym_file.to_bytes().unwrap();
+
+    // Ensure we have exclusive access to mutate __jit_debug_descriptor.
+    let lock = LOCK.lock();
+
+    // Cache the current head of the linked list.
+    let desc = unsafe { ptr::addr_of_mut!(__jit_debug_descriptor) };
+    let old_first = unsafe { (*desc).first_entry };
+
+    // Make a new linked list node.
+    let mut new_ent = Box::new(JitCodeEntry {
+        next_entry: old_first,
+        prev_entry: ptr::null_mut(),
+        symfile_addr: payload.as_ptr() as *const i8,
+        symfile_size: u64::try_from(payload.len()).unwrap(),
+    });
+    let new_ent_ptr = &mut *new_ent as *mut JitCodeEntry;
```

Comment:
It's a good question. One I'm not sure the answer of.

The next and prev fields of the linked list nodes are `*mut` because we absolutely do need to mutate them, and I think that forces us to make `&mut`s here?

Is there a way around this?

Reply:
Have you tried just getting rid of the `&mut`? I *think* `let new_ent_ytr = new_ent as *mut JitCodeEntry` might still let everything compile? If not... you've got to think again I'm afraid. You simply can't have multiple `&mut` references to one thing without being at very high risk of miscompilation due to UB.

Reply:
```
error[E0606]: casting `&JitCodeEntry` as `*mut JitCodeEntry` is invalid
   --> ykrt/src/compile/jitc_yk/gdb.rs:198:23
    |
198 |     let new_ent_ptr = &*new_ent as *mut JitCodeEntry;
```

but:
```
let new_ent_ptr = &*new_ent as *const JitCodeEntry as *mut JitCodeEntry;
```

Works, although it feels like a lie.

Reply:
> Can you not use Box::into_raw?

I could, but then I'd have to store the pointer for reconstitution later. As it stands, freeing of the box is automatic.

Reply:
I was suggesting:

```rust
let new_ent_ytr = new_ent as *mut JitCodeEntry;
```

but you tried 

```rust
let new_ent_ptr = &*new_ent as *mut JitCodeEntry;
```

?

Reply:
@jacob-hughes Good point. `Box::into_raw()` is probably the right thing to do here. It avoids `&mut` entirely. And yes you have to free it manually, but as soon as you've got raw pointers, you're more-or-less guaranteeing to do the moral equivalent of "free" anyway.

Reply:
OK, let's do the into_raw() thing then. Ironically I started doing it that way, but decided against it!

For completeness of discussion:
```
error[E0605]: non-primitive cast: `Box<JitCodeEntry>` as `*mut JitCodeEntry`
   --> ykrt/src/compile/jitc_yk/gdb.rs:198:23
    |
198 |     let new_ent_ptr = new_ent as *mut JitCodeEntry;
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
```

Reply:
> OK, let's do the into_raw() thing then. Ironically I started doing it that way, but decided against it!

Cool. The neat thing about doing it this way is that it removes the `&` references to the box so it's harder to accidentally shoot oneself in the foot.